tests/test_register.py

```diff
@@ -3,11 +3,12 @@
 import pretend
 import pytest
 
-import helpers
 from twine import cli
 from twine import exceptions
 from twine.commands import register
 
+from . import helpers
```

Comment:
Pardon my ignore, but why have we switched from absolute to relative imports here? Was it because of the addition of mypy config to ignore errors in tests? 

I was also looking at https://www.python.org/dev/peps/pep-0008/#imports and it suggests absolute imports over relative imports.

Reply:
Good question; I probably should have called that out. Making `tests` a package (with an `__init__.py`) triggered Python 3's [import behavior](https://portingguide.readthedocs.io/en/latest/imports.html), which treats `import helpers` as a global import (akin to `import twine` or `import requests`). That fails, because `tests/` is not in `sys.path`, which also means `from tests import helpers` would fail. So, I *think* the relative import is the only solution here.

Reply:
_Why_ did we make `tests` a package?

Reply:
Also why do we not want mypy to give us warnings about the types we're passing in to our tests?

Reply:
More indications that I should have been more descriptive in the PR...

> Why do we not want mypy to give us warnings about the types we're passing in to our tests?

While that sounds like a good idea, and something to aspire to, my strong opinion is that it's too much work to do right now, and would add a lot of noise when writing/reading tests, especially related to pytest fixtures and the test doubles (e.g. `pretend`):

```
$ mypy tests
Found 204 errors in 15 files (checked 15 source files)
```

> _Why_ did we make `tests` a package? 

Making `tests` a package enables ignoring errors in `mypy.ini` ([docs](https://mypy.readthedocs.io/en/stable/config_file.html?highlight=wildcard#config-file-format)):

https://github.com/pypa/twine/blob/db6b4703345635ebad2d5fa0d7286ef5a7835ecc/mypy.ini#L53-L54

This mirrors the tox configuration, which only checks `twine`:

https://github.com/pypa/twine/blob/db6b4703345635ebad2d5fa0d7286ef5a7835ecc/tox.ini#L55

And prevents editors that integrate with mypy from showing spurious errors:

<img width="740" alt="Screen Shot 2020-05-11 at 8 50 17 AM" src="https://user-images.githubusercontent.com/1326704/81563612-7e1c0680-9364-11ea-9d1d-8791892d8973.png">

Reply:
So I don't think we've fully considered the ramifications of making tests a package here. I believe that the magical nonsense in our packaging will now treat it as a package. (At least I haven't verified that it won't and I don't trust magical packages when we could instead have explicit configuration like we used to.)

That said, why can't we specify `files` in our config to ignore the tests files? It's documented [here](https://mypy.readthedocs.io/en/stable/config_file.html#import-discovery) and it would allow us to override it on the CLI with posargs. That would have the same value for your editor, I suspect, as what you've done here which requires far more changes than simply updating the config file.

Reply:
I missed (or forgot about) `files`, but unfortunately it doesn't resolve the editor issue, because it executes `mypy tests/test_register.py` on save.

I'm not sure I follow what you mean by "magical nonsense in our packaging".

Re: `tests` being treated as a package, if I understand correctly, the concern is that `tests` would end up in the wheel, and `pip install twine` would then put `tests/*` in `site-packages`. I dug into this a bit, and I'm pretty sure that's not a problem, because `setup.cfg` explicitly defines the packages:

https://github.com/pypa/twine/blob/ae2b876fa20813d979c01fbb341cd12b1d74a47c/setup.cfg#L32-L3

I verified locally that the wheel doesn't contain `tests/*`:

```
$ rm -Rf build dist && python -m pep517.build . > /dev/null && unzip -l dist/*.whl | grep tests
```

Another common pattern is `packages = find:`, which I think is equivalent to `setup(packages=find_packages())`. I verified that using that *would* put `tests/*` in the wheel. However, that can be avoided with options:

```diff
@@ -32,3 +32 @@ classifiers =
-packages =
-    twine
-    twine.commands
+packages = find:
@@ -54,0 +53,3 @@ console_scripts =
+
+[options.packages.find]
+exclude = tests
```

Reply:
> but unfortunately it doesn't resolve the editor issue, because it executes `mypy tests/test_register.py` on save.

That sounds like a defect in the editor, frankly.

> I'm not sure I follow what you mean by "magical nonsense in our packaging".

The team has recently added packaging dependencies that don't pull all of the config from setup.cfg and instead magically divine values. I objected but the changes were merged regardless for an individual's notion of consistency within "their" projects.



> Re: `tests` being treated as a package, if I understand correctly, the concern is that `tests` would end up in the wheel, and `pip install twine` would then put `tests/*` in `site-packages`. I dug into this a bit, and I'm pretty sure that's not a problem, because `setup.cfg` explicitly defines the packages:
> 
> [](https://github.com/pypa/twine/blob/ae2b876fa20813d979c01fbb341cd12b1d74a47c/setup.cfg#L32-L34)

That's satisfactory. We can move forward with this change.

Reply:
> That sounds like a defect in the editor, frankly.

Or at least an opportunity for enhancement: https://github.com/microsoft/vscode-python/issues/11762

> That's satisfactory. We can move forward with this change.

Thanks!